File: app/src/main/java/com/example/a1809zg/welcome/HomeActivity.java
package com.example.a1809zg.welcome;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;

import android.content.ComponentName;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.widget.Toast;

import com.blankj.utilcode.util.LogUtils;
import com.example.a1809zg.MainActivity;
import com.example.a1809zg.R;
import com.example.frame.BaseActivity;
import com.example.frame.CacheMore;
import com.example.frame.FiannceService;
import com.example.net.bean.HomeBean;
import com.example.net.bean.LoginBean;
import com.example.net.bean.UpdataBean;
import com.example.user.AutoService;

public class HomeActivity extends BaseActivity<HomePresenter> implements IHomeView {
private HomePresenter homePresenter;
    private android.widget.TextView miao;
    private android.widget.ProgressBar progressBar;
    private final int ONE_TASK=0;
    private final int ALL_TASK=1;
    private final int DELAY_INDEX=2;

    private final int DELAY=1*1000;
    private boolean HomeFinsh=false;
    private boolean UpdateFinsh=false;
    private boolean AddFinsh=false;
    private int count=3;
    private UpdataBean bean;
    private Intent intent;

    private Handler handler=new Handler(){
        @Override
        public void dispatchMessage(@NonNull Message msg) {
            super.dispatchMessage(msg);
            switch (msg.what){
                case DELAY_INDEX:
                    count--;
                    if (count>0){
                        miao.setText(count+"秒");
                        handler.sendEmptyMessageDelayed(DELAY_INDEX,DELAY);
                    }else {
                        miao.setText(count+"秒");
                        AddFinsh=true;
                        handler.sendEmptyMessage(ONE_TASK);
                    }
                    break;
                case ONE_TASK:
                    if (HomeFinsh&&UpdateFinsh&&AddFinsh){
                        handler.sendEmptyMessage(ALL_TASK);
                    }
                    break;
                case ALL_TASK:

                    Toast.makeText(HomeActivity.this, "1231231", Toast.LENGTH_SHORT).show();
                    all();
                    break;
            }
        }
    };
    @Override
    public void onHomeData(HomeBean homeBean) {
        HomeFinsh = true;
        handler.sendEmptyMessage(ONE_TASK);
        CacheMore.getInstance().setHomeBean(homeBean);

    }

    @Override
    public void onUpdaData(UpdataBean updataBean) {
        Toast.makeText(this, "获取到版本信息："+ updataBean.getResult().getVersion(), Toast.LENGTH_SHORT).show();
        UpdateFinsh=true;
        bean = updataBean;
        handler.sendEmptyMessage(ONE_TASK);
        loadingPage.showSuccessView();
    }


    @Override
    protected int getLayoutId() {
        return R.layout.activity_main2;
    }

    @Override
    protected void initPresenter() {
         homePresenter = new HomePresenter(this);
    }

    @Override
    protected void initData() {
         homePresenter.getHomeData();
         homePresenter.getVersionData();
         handler.sendEmptyMessageDelayed(DELAY_INDEX,DELAY);
        requestPermissions(new String[]{"android.permission.WRITE_EXTERNAL_STORAGE"
                ,"android.permission.READ_EXTERNAL_STORAGE"},100);

        intent = new Intent(this, FiannceService.class);
        startService(intent);

    }

    @Override
    protected void initView() {

        miao = findViewById(R.id.contentTv);
        progressBar = findViewById(R.id.progressBar);
        Intent intent = new Intent(this, AutoService.class);
        startService(intent);
    }

    @Override
    public void showLoaing() {
        loadingPage.showLoadingView();
    }

    @Override
    public void hideLoading() {

    }

    @Override
    public void showError(String msg) {
          Toast.makeText(this, "错误"+msg, Toast.LENGTH_SHORT).show();
    }
    public void all(){
        if (bean!=null){

            int code = bean.getCode();
            if (code==200){
                String version = bean.getResult().getVersion();
                double ver = Double.parseDouble(version);
                Double a = a();
                if (ver>a){
                    AlertDialog.Builder builder = new AlertDialog.Builder(HomeActivity.this);
                    builder.setTitle("版本更新");
                    builder.setPositiveButton("确定", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {

                            bindService(intent, new ServiceConnection() {
                                @Override
                                public void onServiceConnected(ComponentName name, IBinder service) {
                                    LogUtils.d("1233");

                                    FiannceService.FiannceBinder fiannceBinder= (FiannceService.FiannceBinder) service;
                                    FiannceService fiannceService = fiannceBinder.getFiannceService();
                                    fiannceService.DownLoad(bean.getResult().getApkUrl());

                                }

                                @Override
                                public void onServiceDisconnected(ComponentName name) {

                                }
                            },BIND_AUTO_CREATE);

                            Intent intent = new Intent(HomeActivity.this,MainActivity.class);
                            startActivity(intent);
                            finish();
                        }
                    });

                    builder.setNegativeButton("取消", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {
                            startActivity(new Intent(HomeActivity.this, MainActivity.class));
                            dialog.dismiss();
                        }
                    });
                    AlertDialog alertDialog = builder.create();
                    alertDialog.show();
                }
            }
        }
    }
    public Double a(){
        PackageManager packageManager = getPackageManager();
        PackageInfo packageInfo =null;
        try {
            packageInfo = packageManager.getPackageInfo(getPackageName(), 0);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        String versionName = packageInfo.versionName;
        double num = Double.parseDouble(versionName);
        return num;
    }

    @Override
    public void destroy() {
        super.destroy();
        handler.removeCallbacksAndMessages(true);

        homePresenter.detachView();

    }

    @Override
    public void onConnect() {
         Toast.makeText(this, "有网络了", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onDisConnect() {
        Toast.makeText(this, "网络错误", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onLoginChange(LoginBean loginBean) {

    }
}

File: app/src/main/java/com/example/financial/ui/home/HomeActivity.kt
package com.example.financial.ui.home

import android.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.os.Handler
import android.os.Message
import android.util.Log
import android.widget.Toast
import com.example.financial.MainActivity
import com.example.financial.R
import com.example.financial.base.Product
import com.example.financial.base.ResultX
import com.example.financial.ui.product.ProductFragment
import com.example.financial.ui.product.ProductModle
import com.example.financial.ui.product.ProductPresenter
import com.example.financial.ulit.DataUlit
import com.example.frame_library.mvp.BaseActitvty
import kotlinx.android.synthetic.main.home_activity.*
import java.util.*
import kotlin.concurrent.timerTask

class HomeActivity:BaseActitvty<HomePresenter>(),HomeCanter.View {
    private val LODINDATA:Int=-10001
    private val UPDATA:Int=-10002
    private val TIME:Int=-10003
    private val handler:Handler=object :Handler(){
        private var ISLODINDATA:Boolean=false
        private var ISUPDATA:Boolean=false
        private var ISTIME:Boolean=false
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)

            when(msg.what){
                LODINDATA->ISLODINDATA=true
                UPDATA->ISUPDATA=true
                TIME->ISTIME=true
            }

            if (ISLODINDATA&&ISTIME&&ISUPDATA){
                UpData()
            }
        }
    }

    /***
     *三秒倒计时
     */
    private val timer:Timer by lazy {
        var time:Int =3
        var timer1 = Timer()
        timer1.schedule(timerTask {
            if (time == 0) {
                handler.sendEmptyMessage(TIME)
                cancel()
            }else{
            handler.post {
                act_home_time_text.setText("${time}")
            }
            time--
            }

        }, 1000, 1000)
        timer1
    }

    override fun bandLayoutId(): Int = R.layout.home_activity

    override fun initView() {

    }

    override fun initData() {
        act_home_version_text.setText(getVersionName()+"")

        mPresenter!!.getVersion()

        timer

        mPresenter.LoginData()
    }

    /**
     *获取当前版本名称
     *  @return 以String形式返回版本名称
    */
     fun getVersionName(): String? {
        return packageManager.getPackageInfo(packageName,0)!!.versionName
    }

    /**
     * 获取当前版本号
     * @return Int形式当前版本号
     * */
    fun getVersionNumer():Int{
        return packageManager.getPackageInfo(packageName,0)!!.versionCode
    }
    protected var IsUp:Boolean=false
    /**
     * 从网络获取最新版本号，判断
     * */
    override fun IsUpdate(versionCode: Int) {
        handler.sendEmptyMessage(UPDATA)
        IsUp=versionCode==getVersionNumer()
    }

    override fun onLodinData(product: Product<ResultX>) {
        handler.sendEmptyMessage(LODINDATA)
        DataUlit.product=product
    }

    private fun UpData(){
        if (!IsUp){
            var create = AlertDialog.Builder(this)
                .setIcon(R.drawable.ic_launcher_background)
                .setMessage("优化bug,新增功能")
                .setTitle("下载最新版本")
                .setNegativeButton("取消", object : DialogInterface.OnClickListener {
                    override fun onClick(dialog: DialogInterface?, which: Int) {
                        startActivity(Intent(this@HomeActivity,MainActivity::class.java))
                        finish()
                    }
                })
                .setPositiveButton("确定",object : DialogInterface.OnClickListener {
                    override fun onClick(dialog: DialogInterface?, which: Int) {

                    }
                })
                .create()
            create.show()
        }
    }

    /***
     * @return初始化Presneter
     */
    override fun setPresneter(): HomePresenter=HomePresenter(HomeModle(), this)

    override fun onDestroy() {
        super.onDestroy()
        timer.cancel()
    }
}